File: neetcode/7. trees/1.binaryTreeInorderTraversal.cpp
// Given the root of a binary tree, return the inorder traversal of its nodes' values.

# include <vector>
# include <stack>
using namespace std;

// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode* right) : val(x), left(left), right(right) {}
};

// Recursive solution
class Solution {
public:
  vector<int> inorderTraversal(TreeNode* root) {
    vector<int> result;
    inorder(root, result);
    return result;
  }
private:
  void inorder(TreeNode* root, vector<int>& result) {
    if (root == nullptr) return;
    inorder(root->left, result);
    result.push_back(root->val);
    inorder(root->right, result);
  }
};

// Iterative solution
class Solution {
public:
  vector<int> inorderTraversal(TreeNode* root) {
    vector<int> res;
    stack<TreeNode*> s;
    TreeNode* curr = root;
    while (curr != nullptr || !s.empty()) {
      while (curr != nullptr) {
        s.push(curr);
        curr = curr->left;
      }
      curr = s.top();
      s.pop();
      res.push_back(curr->val);
      curr = curr->right;
    }
    return res;
  }
};

File: neetcode/7. trees/1.binaryTreeInorderTraversal.py
# Given the root of a binary tree, return the inorder traversal of its nodes' values.

from typing import List

# Definition for a binary tree node.
class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right

# Recursive algorithm: time complexity: O(n)
class Solution:
  def inorderTraversal(self, root: TreeNode) -> List[int]:
    res = []

    def inorder(root):
      if not root:
        return
      inorder(root.left)
      res.append(root.val)
      inorder(root.right)
    
    inorder(root)
    return res

# Iterative algorithm: time complexity: O(n)
class Solution:
  def inorderTraversal(self, root: TreeNode) -> List[int]:
    res = []
    stack = []
    while root or stack:
      while root:
        stack.append(root)
        root = root.left
      root = stack.pop()
      res.append(root.val)
      root = root.right
    return res